skill-tree/src/main/java/com/RDS/skilltree/services/EndorsementServiceImplementation.java

```diff
@@ -43,59 +69,63 @@ public EndorsementViewModel create(CreateEndorsementViewModel endorsementViewMod
-        Optional<UserModel> endorseDetails = userRepository.findById(endorseId);
-        Optional<UserModel> endorserDetails = userRepository.findById(endorserId);
-
-        if (endorserDetails.isEmpty()) {
-            throw new UserNotFoundException("Cannot create endorsement for the current user");
-        }
-
-        if (endorseDetails.isEmpty()) {
-            throw new UserNotFoundException("Endorse not found");
-        }
 
         if (skillDetails.isEmpty()) {
-            throw new SkillNotFoundException(String.format("Skill id: %s not found", skillId));
+            log.error(String.format("Skill id: %s not found", skillId));
+            throw new SkillNotFoundException("Skill does not exist");
         }
 
-        List<UserSkillsModel> userSkillEntry =
+        // Get endorse(person being endorsed) & endorser(person endorsing) details from RDS
+        RdsGetUserDetailsResDto endorseDetails = rdsService.getUserDetails(endorseId);
+        RdsGetUserDetailsResDto endorserDetails = rdsService.getUserDetails(endorserId);
+
+        List<UserSkills> userSkillEntry =
                 userSkillRepository.findByUserIdAndSkillId(endorseId, skillId);
-        Endorsement endorsement = new Endorsement();
 
-        endorsement.setMessage(message);
-        endorsement.setSkill(skillDetails.get());
-        endorsement.setEndorse(endorseDetails.get());
-        endorsement.setEndorser(endorserDetails.get());
+        Endorsement endorsement =
+                Endorsement.builder()
+                        .skill(skillDetails.get())
+                        .message(message)
+                        .endorseId(endorseId)
+                        .endorserId(endorserId)
+                        .build();
 
+        // If a skill request is not created then create one
+        // This is because there is no specific api to create a skill request at the time of writing
         if (userSkillEntry.isEmpty()) {
-            UserSkillsModel userSkillsModel = new UserSkillsModel();
-            userSkillsModel.setUser(endorseDetails.get());
-            userSkillsModel.setSkill(skillDetails.get());
-            userSkillsModel.setStatus(UserSkillStatusEnum.PENDING);
+            UserSkills userSkills =
+                    UserSkills.builder().userId(endorseId).skill(skillDetails.get()).build();
 
-            userSkillRepository.save(userSkillsModel);
+            userSkillRepository.save(userSkills);
         }
 
         Endorsement newEndorsement = endorsementRepository.save(endorsement);
 
-        return EndorsementViewModel.toViewModel(newEndorsement);
+        return EndorsementViewModel.toViewModel(
+                newEndorsement,
+                UserViewModel.toViewModel(endorseDetails.getUser()),
+                UserViewModel.toViewModel(endorserDetails.getUser()));
     }
 
     @Override
     public EndorsementViewModel update(Integer endorsementId, UpdateEndorsementViewModel body) {
```

Comment:
similar Similar to the getAllEndorsementsBySkillId function, there are separate calls to rdsService.getUserDetails for both endorseId and endorserId within the update function.  Here, you can potentially optimize by following one of the approaches mentioned previously:

Combining calls based on userIds: If the rdsService supports fetching multiple users in a single call, you could potentially combine fetching user details for endorseId and endorserId into one API call.

Checking for duplicates before fetching:  An alternative is to check if the user details already exist in a local variable before making the API call. You can create a temporary map to store the fetched user details within this function and use it to avoid redundant calls.

```
public EndorsementViewModel update(Integer endorsementId, UpdateEndorsementViewModel body) {
    Optional<Endorsement> exitingEndorsement = endorsementRepository.findById(endorsementId);

    if (exitingEndorsement.isEmpty()) {
        log.error(String.format("Endorsement with id: %s not found", endorsementId));
        throw new EndorsementNotFoundException("Endorsement not found");
    }

    Endorsement endorsement = exitingEndorsement.get();
    // ... update endorsement logic ...

    Endorsement savedEndorsementDetails = endorsementRepository.save(endorsement);

    // Store fetched users to avoid redundant calls
    Map<String, UserViewModel> userDetails = new HashMap<>();
    RdsGetUserDetailsResDto endorseDetails;
    if (!userDetails.containsKey(savedEndorsementDetails.getEndorseId())) {
        endorseDetails = rdsService.getUserDetails(savedEndorsementDetails.getEndorseId());
        userDetails.put(endorseDetails.getUser().getUserId(), UserViewModel.toViewModel(endorseDetails.getUser()));
    } else {
        endorseDetails = null;
    }

    RdsGetUserDetailsResDto endorserDetails;
    if (!userDetails.containsKey(savedEndorsementDetails.getEndorserId())) {
        endorserDetails = rdsService.getUserDetails(savedEndorsementDetails.getEndorserId());
        userDetails.put(endorserDetails.getUser().getUserId(), UserViewModel.toViewModel(endorserDetails.getUser()));
    } else {
        endorserDetails = null;
    }

    return EndorsementViewModel.toViewModel(
            savedEndorsementDetails,
            userDetails.get(savedEndorsementDetails.getEndorseId() != null ? savedEndorsementDetails.getEndorseId() : (endorseDetails != null ? endorseDetails.getUser().getUserId() : null)),
            userDetails.get(savedEndorsementDetails.getEndorserId() != null ? savedEndorsementDetails.getEndorserId() : (endorserDetails != null ? endorserDetails.getUser().getUserId() : null)));
}

```